consensus/types/src/shuffling_id.rs

```diff
@@ -0,0 +1,64 @@
+use crate::*;
+use serde_derive::{Deserialize, Serialize};
+use ssz_derive::{Decode, Encode};
+use std::hash::Hash;
+
+/// Can be used to key (ID) the shuffling in some chain, in some epoch.
+///
+/// ## Reasoning
+///
+/// We say that the ID of some shuffling is always equal to a 2-tuple:
+///
+/// - The epoch for which the shuffling should be effective.
+/// - A block root, where this is the root at the *last* slot of the penultimate epoch. I.e., the
+/// final block which contributed a randao reveal to the seed for the shuffling.
+///
+/// The struct stores exactly that 2-tuple.
+#[derive(Debug, PartialEq, Clone, Hash, Serialize, Deserialize, Encode, Decode)]
+pub struct ShufflingId {
+    pub shuffling_epoch: Epoch,
+    shuffling_decision_block: Hash256,
+}
+
+impl ShufflingId {
+    /// Using the given `state`, return the shuffling id for the shuffling at the given
+    /// `relative_epoch`.
+    ///
+    /// The `block_root` provided should be either:
+    ///
+    /// - The root of the block which produced this state.
+    /// - If the state is from a skip slot, the root of the latest block in that state.
+    pub fn new<E: EthSpec>(
+        block_root: Hash256,
+        state: &BeaconState<E>,
+        relative_epoch: RelativeEpoch,
+    ) -> Result<Self, BeaconStateError> {
+        let shuffling_epoch = relative_epoch.into_epoch(state.current_epoch());
+
+        // Taking advantage of saturating subtraction on slot and epoch.
+        //
+        // This is the final slot of the penultimate epoch.
+        let shuffling_decision_slot =
+            (state.current_epoch() - 1).start_slot(E::slots_per_epoch()) - 1;
```

Comment:
Shouldn't this subtract from the `shuffling_epoch` in question instead of the current epoch? I.e.

```rust
(shuffling_epoch - 1).start_slot(E::slots_per_epoch()) - 1
```

Or even clearer:

```rust
(shuffling_epoch - 2).end_slot(E::slots_per_epoch())
```

We could also replace the 2 by `1 + spec.min_seed_lookahead`, to be robust against potential spec updates (unlikely, but possible).

Reply:
You're right about the `shuffling_epoch`, thanks!

> `(shuffling_epoch - 2).end_slot(E::slots_per_epoch())`

Unfortunately this doesn't work for cases like `state.slot ==1`, where you'll end up with a shuffling decision slot of `31`.

Reply:
Ah, yeah!